language/move-vm/runtime/src/runtime.rs

```diff
@@ -73,14 +74,32 @@ impl VMRuntime {
             .finish(Location::Undefined));
         }
 
-        // Make sure that there is not already a module with this name published
-        // under the transaction sender's account.
         let module_id = compiled_module.self_id();
+
+        // For now, we assume that all modules can be republished, as long as the new module is
+        // backward compatible with the old module.
+        //
+        // TODO: in the future, we may want to add restrictions on module republishing, possibly by
+        // changing the bytecode format to include an `is_upgradable` flag in the CompiledModule.
         if data_store.exists_module(&module_id)? {
-            return Err(
-                PartialVMError::new(StatusCode::DUPLICATE_MODULE_NAME).finish(Location::Undefined)
-            );
-        };
+            let old_module_bytes = data_store.load_module(&module_id)?;
+            let old_module = match CompiledModule::deserialize(&old_module_bytes) {
+                Ok(module) => module,
+                Err(err) => {
+                    warn!(*log_context, "[VM] module deserialization failed {:?}", err);
+                    return Err(err.finish(Location::Undefined));
+                }
+            };
+            let old_m = normalized::Module::new(&old_module);
+            let new_m = normalized::Module::new(&compiled_module);
+            let compat = Compatibility::check(&old_m, &new_m);
+            if !compat.is_fully_compatible() {
+                return Err(
+                    PartialVMError::new(StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE)
+                        .finish(Location::Undefined),
+                );
+            }
+        }
```

Comment:
As I mentioned above, I don't think this is safe. You previous version (boolean flag in the inner layer + separate interfaces for publishing and republishing in the outer layer) seems more favorable to me.

Reply:
Curious: do you think there should be separate transaction types for publish and republish? I was thinking that these should be the same transaction type and same VM API because (1) they require exactly the same data/permissions and (2) I'm not sure what separating them buys us. We have tried fairly hard to keep the transaction types and VM public APIs minimal, so I think we should only add new surface area if there is a strong motivation.

Reply:
Ah I see where our disagreement comes from. To me
(1) publishing and republishing require totally different permissions -- republishing is changing an already-agreed contract and thus unsafe and at least in the near future, require association approval. Normal publishing isn't going to break anything so you just need to prove you are the owner of the account (and that you are allowed to publish onto the network).
(2) Another benefit the separation might buy us is reduced chance to make a mistake -- republishing is a lot more risky and it sounds like a bad thing to me you could accidentally trigger this when only meant to publish a new module.

Reply:
> republishing is changing an already-agreed contract and thus unsafe and at least in the near future, require association approval
> Normal publishing isn't going to break anything so you just need to prove you are the owner of the account (and that you are allowed to publish onto the network).

Today, only the Association can publish and change modules. Letting another party either publish new modules or update new modules are equally risky in my book. For example: the new module may implement a use-case we do not want or are not legally allowed to have, and an updrage to an existing module has the same issue.

So I do not agree that these necessarily require different permissions, but I do agree that once we allow non-Association parties to publish, we must think carefully about what the rules are. I would suggest something like: we can gate module publishing by Association approval, but the value of the upgradability byte is one of the criteria considered in the approval process (e.g., it's easier to get an immutable module approved than one that allows arbitrary upgrades).

However, let's not get too far ahead of ourselves here: this PR is just allowing the Association to republish its own modules, which I think is a very safe and reasonable place to start.
> it sounds like a bad thing to me you could accidentally trigger one when you only meant to publish a new module.

I think that this is quite unlikely because of the backward compatibility checks. Specifically, in order to do this, you new module would need to coincidentally:
1. have the same name as the old one
2. be published at the same address as the old one
3. contain a superset of the structs of the old one
4. contain a superset of the public functions of the old one

If (1) - (4) were all true, I would question why the author was trying to write a new module in the first place instead of extending the existing one :).

Reply:
Let's continue the discussion in this thread :)

I get your points. I agree that for the Diem side of things, since we currently impose the strictest access control for both publishing and republishing, the overwriting behavior will likely be fine and there's no need to create a new transaction type just for the sake of minimizing risk in the short term. And as you mentioned, if one day we open up publishing to partners or even the public, things will need to be revisited.

I now realize that my "opposition" was more of a matter of Move VM API design: it felt weird to me that the VM is making the choice for its clients, always allowing modules to be updated silently. I feel like this choice should be made by the adapter and the VM's role is to make options available. To me the VM's interfaces should be expressive and somewhat generic, and should not assume the client works in a particular way. I think we're now in a weird time when Diem is still the single monolithic client of Move, but other clients are also emerging, hence the confusion over how far in the future we're looking into. Anyways, let's compare the options we have:
- (current impl) only `fn publish` (overwriting)
- `fn publish(..., allow_republishing: bool)` (error when module exists and flag set to false; overwrite if otherwise)
- `fn publish(...)` (error when module exists) + `fn update(...)` (publish the module if one doesn't exist) -- equivalent to above, just that we now have two entries instead of a boolean flag
- `fn publish(...)` (error when module exists) + `fn update(...)` (error when module doesn't exist)

> I think that this is quite unlikely because of the backward compatibility checks.

This is a very good point and I think following this logic, we can say that it's not well justified that the reason why the VM would have separate entries or the boolean flag is because it wants to protect its clients. Still, we need to consider whether the VM should make these options available to the clients.

Reply:
That is a nice summary of options!

But the only reason I can see for having multiple entry points (i.e., `fn publish`, `fn update`, and even the `fn publish(..., allow_republish: bool)`) in the VM is to support different transaction types. So the underlying question behind these options is: shall we have different transaction types for:

- Publish_NoRepublishing
- Publish_AllowRepublishingWhenCompatible
- Publish_UpdateOnly

And how do we encode that in the transaction types...

Reply:
>  To me the VM's interfaces should be expressive and somewhat generic, and should not assume the client works in a particular way. 

I agree with this principle, and think that a single `publish(M)` API for both publishing and updates is in keeping with it.

Note that this minimal API can support all three transaction types mentioned by Meng. Keep in mind that an adapter has access to the global storage, which gives it a *lot* of flexibility in how it uses the VM APIs. For example: a client adapter can implement a `PublishAllowRepublishing(M)` transaction type as `if  M does not exists in global storage call VM::publish(M) else err` and a `PublishUpdateOnly(M)` transaction type as `if M exists in global storage call VM::publish(M) else err`, and an `PublishAllowRepublishingWhenCompatible(M)` transaction type as `call VM::publish(M)`.

Now, to answer Meng's question about which transaction types we want in Diem: I think the single module publish transaction type we have is just fine for now unless we find a strong motivation for the other types. We are already considering some changes to module publishing (e.g., multiple modules at a time, multiple signers), and that will be more painful if we duplicate publishing functionality across multiple tx types.

Reply:
> a PublishUpdateOnly(M) transaction type as if M exists in global storage call VM::publish(M) else err
This makes me feel a lot better about the current solution! And I agree that the module publishing APIs will likely go through more changes in the incoming days, so let's not worry too much about the detailed differences here.

Reply:
I can provide some Move module republish/upgrade practical experience, maybe can help this discussion.

We use MoveVM in starcoin project, and starcoin supports Module upgrade.

1. We remove the repeat module limit in move_vm [runtime.rs](https://github.com/starcoinorg/diem/blob/starcoin_01_05/language/move-vm/runtime/src/runtime.rs)
2. Implement compatibility check in starcoin move VM(https://github.com/starcoinorg/starcoin/blob/master/vm/vm-runtime/src/starcoin_vm.rs#L274).
3. Define [ModuleUpgradeStrategy](https://github.com/starcoinorg/starcoin/blob/master/vm/stdlib/modules/PackageTxnManager.move#L52 ) in stdlib,  module publisher can write an onchain config resource to choose a ModuleUpgradeStragety for all module at the publisher's address.
5. The transaction payload supports a [Package](https://github.com/starcoinorg/starcoin/blob/master/vm/types/src/transaction/mod.rs#L245) type, it contains a Vec<Module> and an init Script(for genesis transaction).
6. When executing the Package transaction, VM checks the Compatibility and UpgradeStragety if the module already exists.

So,  I have the following suggestions:

1. Provider a VM config that allows developers to configure the behavior of the VM, such as check compatibility,  allow publish, enable debug, etc.
2. The VM public API keep simple, do not care about Republish/Compatible, let the upper layer applications to decide how to use the Move VM, such as in Dev env, it allows republish module without check Compatibility.

Reply:
@jolestar Thanks for sharing your experience and advice!

> We help Move lang support an approach to marker a method is deprecated.

Sounds interesting--can you say more about how this part works?

Reply:
@jolestar Thank you too for the sharing and advice!

I am also curious about this part:
> When executing the Package transaction, VM checks the Compatibility and UpgradeStragety if the module already exists.

Does this implies that the VM knows where the `UpgradeStragety` is stored in the account address? Well, since this `UpgradeStragety` is a resource, the VM should know, but just want to confirm that whether the VM is hardcoded to look for a `UpgradeStragety` resource when doing the `Package` transaction.

Reply:
> @jolestar Thanks for sharing your experience and advice!
> 
> > We help Move lang support an approach to marker a method is deprecated.
> 
> Sounds interesting--can you say more about how this part works?

typo s/help/hope/ ， expect Move lang support @Deprecated annotation like java or rust.

Reply:
> @jolestar Thank you too for the sharing and advice!
> 
> I am also curious about this part:
> 
> > When executing the Package transaction, VM checks the Compatibility and UpgradeStragety if the module already exists.
> 
> Does this implies that the VM knows where the `UpgradeStragety` is stored in the account address? Well, since this `UpgradeStragety` is a resource, the VM should know, but just want to confirm that whether the VM is hardcoded to look for a `UpgradeStragety` resource when doing the `Package` transaction.

VM needs to read on chain state from state store, like VMConfig, if No UpgradeStragety in the Package's module address, use default strategy.